teuthology/provision.py

```diff
@@ -29,105 +30,200 @@ def _get_downburst_exec():
-        has_config = hasattr(ctx, 'config') and ctx.config is not None
-        if has_config and 'downburst' in ctx.config:
-            log.warning(
-                'Usage of a custom downburst config has been deprecated.'
-            )
+    def create(self):
+        """
+        Launch a virtual machine instance.
 
+        If creation fails because an instance with the specified name is
+        already running, first destroy it, then try again. This process will
+        repeat two more times, waiting 60s between tries, before giving up.
+        """
+        if not self.executable:
+            log.error("No downburst executable found.")
+            return False
+        self.build_config()
+        success = None
+        with safe_while(sleep=60, tries=3,
+                        action="downburst create") as proceed:
+            while proceed():
+                (returncode, stdout, stderr) = self._run_create()
+                if returncode == 0:
+                    log.info("Downburst created %s: %s" % (self.name,
+                                                           stdout.strip()))
+                    success = True
+                    break
+                elif stderr:
+                    # If the guest already exists first destroy then re-create:
+                    if 'exists' in stderr:
+                        success = False
+                        log.info("Guest files exist. Re-creating guest: %s" %
+                                 (self.name))
+                        self.destroy()
+                    else:
+                        success = False
+                        log.info("Downburst failed on %s: %s" % (
+                            self.name, stderr.strip()))
+                        break
+            return success
+
+    def _run_create(self):
+        """
+        Used by create(), this method is what actually calls downburst when
+        creating a virtual machine instance.
+        """
+        if not self.config_path:
+            raise ValueError("I need a config_path!")
+        shortname = decanonicalize_hostname(self.name)
+
+        args = [self.executable, '-c', self.host, 'create',
+                '--meta-data=%s' % self.config_path, shortname,
+                ]
         log.info("Provisioning a {distro} {distroversion} vps".format(
-            distro=os_type,
-            distroversion=os_version
+            distro=self.os_type,
+            distroversion=self.os_version
         ))
+        proc = subprocess.Popen(args, stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+        out, err = proc.communicate()
+        return (proc.returncode, out, err)
+
+    def destroy(self):
+        """
+        Destroy (shutdown and delete) a virtual machine instance.
+        """
+        executable = self.executable
+        if not executable:
+            log.error("No downburst executable found.")
+            return False
```

Comment:
I didn't see tests for this, shouldn't be difficult to add one for this condition